### PR TITLE
Add convex hull trick draft

### PR DESCRIPTION
**Pull request by aftermaaath**
- [x] Read [Contribute to NTHU-CPP](https://nthu-cp.github.io/NTHU-CPP/others/contribution.html).
- [x] Rebase to the latest main branch.
- [x] List all the references.
- [x] Successfully build the website by mdBook with no error after the change.
- [x] Exclude any irrelevant files.
- [x] Link to the issue which is related to your change.
- [x] Review the content by myself.
- [x] Pass tests/CI.

### CODE REVIEW

**Comment by aftermaaath** [comment]
I would fix the workflow right after I put all solutions for exercises. The outline and most of the content is finished, except for the images, the solutions & problem descriptions for the exercises, and the summary. I think I can ask for comments first so I can fix the content earlier. Thanks.

@harry900831

File: src/dp/convex_hull_optimization.md [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+
+至於二分搜要搜什麼呢？注意到我們維護的是一個凸包，凸包上的點的 \\(x\\) 座標是單調的。而每當我們要查詢的時候，我們想找的那條線上的兩個頂點會在當前查詢的 \\(x\\) 座標的左右兩側：
+
+// 一個凸包的圖，標記兩側頂點
+
+因此我們在二分搜時，可以比較當前線段與下一條線段交點的 \\(x\\) 座標，以及當前要查詢的 \\(x\\) 座標，若前者較大則右界改為當前 \\(mid\\)、反之更改左界。
+
+將查詢改為二分搜會讓複雜度多一個 \\(\log\\)，總複雜度是 \\(O(n\log n)\\)。
+
+以下是用此方法解 [CSES - Monster Game I](https://cses.fi/problemset/task/2084/) 的範例 code。
```
[the remark this answers is not in the document]
Comment (harry900831) [reply src/dp/convex_hull_optimization.md:161]: 同時 CSES - monster Game I 的轉移是有點太簡單，可以的話看是要不要就這個新題 或是額外再這邊再講解一個要推一下子才能推出 ax+b 的形式的題目。

File: src/dp/convex_hull_optimization.md [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+斜率不單調的情況更為複雜，我們不能再使用 deque 或是 set 來維護凸包，以下會介紹幾個解決此問題的方法。
+
+#### CDQ 分治
+
+直接來看一個題目。
+> [CSES - Monster Game II](https://cses.fi/problemset/task/2085)
+>
+> 有 \\(n\\) 個關卡，每關有一個怪物，通關時必須從第一關開始依序通關。打敗一關的怪物需消耗 \\(sf\\) 的 cost，\\(s\\) 為怪物血量，\\(f\\) 為能力值。你會有一個起始的能力值，每到一關，你可以選擇是否打敗該關怪物，並將能力值更新為 \\(f_i\\)。求打敗第 \\(n\\) 關怪物的最小 cost。
+
+題目大致與 [CSES - Monster Game I](https://cses.fi/problemset/task/2084/) 相同，但此題不再保證 \\(f\\) 與 \\(s\\) 的單調性，意即斜率與查詢都不再單調。不過透過 CDQ 分治，我們可以「人為」創造出斜率的單調性。
```
Comment (harry900831) [inline src/dp/convex_hull_optimization.md:255]: 一樣需要推一下遞迴式

File: src/dp/convex_hull_optimization.md [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+
+我們一樣要在遞迴過程中，將序列分左半邊與右半邊處理。前面提到我們想創造出斜率的單調性，因此在遞迴過程中我們要維護一個斜率單調的序列，且裡面沒有不會被取到的直線。先不論直線如何計算，討論維護的部分。維護的方式類似 merge sort 的過程，只是要注意新增直線到序列尾端時，一樣要做 pop 的動作，將不可能會被取到的直線刪除。
```
Comment (harry900831) [inline src/dp/convex_hull_optimization.md:257]: 要遞迴什麼？ 這部分需要更詳細說明，我看過去看不太懂。

File: src/dp/convex_hull_optimization.md [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+    cout << dp[n] << "\n";
+    return 0;
+}
+```
+
+</details>
+
+#### 動態凸包
+
+// 希望這一part可以日後再補Q
```
Comment (harry900831) [inline src/dp/convex_hull_optimization.md:523]: 我其實不太懂這裡的動態凸包是指什麼，不過應該要提到維護插入線段查詢x的最高點是有模板可用的，可參考：https://robert1003.github.io/2020/02/17/dp-opt-convex-hull-trick.html#convex-hull-trick-cht


大家的codebook大概都會備一份

Comment (harry900831) [reply src/dp/convex_hull_optimization.md:523]: 可以的話底下可以挑個別人的模板或自己寫一個，然後後面的 excercise 就依靠那個模板來解。

**Comment by harry900831** [comment]
Furthermore, you could provide some comments about the references about which on is recommended to read, and hint the reader which resources is good for them if they want to learn more.

File: src/dp/convex_hull_optimization.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (aftermaaath) [reply src/dp/convex_hull_optimization.md:161]: 查詢不單調的範例，我原本也想找個新題，找很久找不到才先參考其他 tutorial 這樣寫Q，我找到題目後會將這部分更新。

Exercise 的題目會有更複雜的轉移，這樣還需要再講解一個嗎？（或是我可以直接搬上來？exercise 也會有題解）

[the remark this answers is not in the document]
Comment (harry900831) [reply src/dp/convex_hull_optimization.md:161]: 放在這部分的題目都需要「詳細」的題解，包含轉移是怎麼推以及實作細節，因此我認為在這邊需要再放一題難一點的。

Exercise 的目的是提供讀者練習的題目，因此他的解釋不需要太完整，除非是網路上都找不到解答而他又很難的話。

**Review by harry900831 — CHANGES_REQUESTED** [review]
- The content overall LGTM!
- Please refactor your code
- I might merge it after your revision so please check the whole article again by yourself
- Please re-request review after you finish it

Thanks

File: src/dp/convex_hull_optimization.md [hunk 5]
```diff
@@ -0,0 +1,976 @@
...
+>
+> - \\(1\leq n\leq 2\times 10^5\\)
+
+令 \\(f(i)\\) 為殺掉第 \\(i\\) 關怪物所需花的最少時間。
+
+轉移式為：\\(F(i)=\underset{i>j}{\min}(F(j)+f_j\times s_i)\\)
+
+題目大致與 [CSES - Monster Game I](https://cses.fi/problemset/task/2084/) 相同，但此題不再保證 \\(f\\) 與 \\(s\\) 的單調性，意即斜率與查詢都不再單調。不過透過 CDQ 分治，我們可以「人為」創造出斜率的單調性。
+
+CDQ 分治的過程就是對於一個序列 \\([l, r]\\)，我們先遞迴解決 \\([l, mid]\\) 的子問題，再處理 \\([l, mid]\\) 對 \\([mid + 1, r]\\) 的貢獻，最後再遞迴處理 \\([mid + 1, r]\\)。
```
[the remark this answers is not in the document]
Comment (harry900831) [reply src/dp/convex_hull_optimization.md:470]: 我先在覺得斜率優化用CDQ分治的教學好像應該在 #63 裡面詳細講才對，但是你寫的整體來說還不錯，看要不要修改一下然後整塊搬運過去給 #63。

@chyyen @aftermaaath 你們有什麼想法嗎？

[the remark this answers is not in the document]
Comment (aftermaaath) [reply src/dp/convex_hull_optimization.md:470]: Hmm 我其實覺得都可，當初是參考 OI wiki 才寫在這邊，可能看看 CDQ 那裡的架構如何，我沒有太大的意見。

[the remark this answers is not in the document]
Comment (chyyen) [reply src/dp/convex_hull_optimization.md:470]: 我也覺得 ok

[the remark this answers is not in the document]
Comment (harry900831) [reply src/dp/convex_hull_optimization.md:465]: 所以這部分就交給 @chyyen 寫囉

File: src/dp/convex_hull_optimization.md [hunk 6]
```diff
@@ -0,0 +1,976 @@
...
+
+    <img src="image/convex_hull_optimization/tioj1676_lines_4.png" width="500" style="display:block; margin: 0 auto;"/>
+
+    假設這時最左邊的線過期了，於是我們將它移除：
+
+    <img src="image/convex_hull_optimization/tioj1676_lines_5.png" width="500" style="display:block; margin: 0 auto;"/>
+
+    這時我們發現其實剛剛被我們移除掉的藍色線才能提供最好答案，但我們已經將它移除了。從這個例子我們發現，在加了直線會過期的限制後，我們除了要在直線過期時移除它之外，我們也要更改一下移除不在凸包上直線的策略。
+
+    注意到在綠色線被移除之前，藍色線都不可能提供答案。因此我們要多判斷的是藍色線在綠色線被移除之後，會不會成為可能的轉移來源。具體而言，在我們放入紅色線前，要決定藍色線是否要被移除，而方法是多判斷在綠色線被 pop 時的那一個查詢，藍色線會不會提供比紅色線更好的答案。
```
[the remark this answers is not in the document]
Comment (aftermaaath) [reply src/dp/convex_hull_optimization.md:239]: I think our approach is the same, and also I think it is correct (not proved by AC). Should I mention this link still?

File: src/dp/convex_hull_optimization.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (harry900831) [reply src/dp/convex_hull_optimization.md:465]: 可以一句話提一下 可以使用 CDQ 分治

File: src/dp/convex_hull_optimization.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (harry900831) [reply src/dp/convex_hull_optimization.md:239]: If the approach is the same, then never mind.

File: src/dp/convex_hull_optimization.md [hunk 7]
```diff
@@ -0,0 +1,1055 @@
...
+>
+> 有 \\(p\\) 個 feeders 想把 \\(m\\) 隻貓接回農場。貓現在分布在各個山上，共有 \\(n\\) 座山，由左至右編號 \\(1\sim n\\)，相鄰的兩座山 \\(i\\) 與 \\(i-1\\) 山隔著 \\(d_i\\) 公尺。每隻貓都會給定其在哪座山上，以及一個時間 \\(t_i\\)，表示經過 \\(t_i\\) 單位時間後貓會開始等待被接走。現在 feeders 要從 \\(1\\) 號山出發往右走，以每單位時間一公尺的速度去接貓，當 feeder 抵達一座山，他會接走該座山中所有正在等待的貓。現在我們要安排 feeders 的出發時間，使得所有貓的等待時間總和最短。
+>
+> - \\(2 \leq n \leq 10^5\\)
+> - \\(1 \leq m \leq 10^5\\)
+> - \\(1 \leq p \leq 10^2\\)
+
+<details><summary> Solution </summary>
+
+先將所有貓的 \\(t_i\\) 減掉 \\(h_1\\) 至 \\(h_i\\) 的距離，這樣就可以得到每隻貓最早能被接走時，feeder 的出發時間。
```
Comment (fffelix-huang) [inline src/dp/convex_hull_optimization.md:918]: 上面沒有提到 $h_i$

**Review by fffelix-huang — CHANGES_REQUESTED** [review]
Great content!
- Please fix minor errors
- Fix textlint
- Re-request review after you finish it
Thanks~

File: src/SUMMARY.md [hunk 8]
```diff
@@ -11,6 +11,7 @@
 - [Greedy](greedy/intro.md)
   - [Basic Greedy](greedy/basic.md)
 - [Dynamic Programming]()
+  - [Convex hull optimization](dp/convex_hull_optimization.md)
```
Comment (harry900831) [inline src/SUMMARY.md:14]: Convex Hull Optimization

**Review by harry900831 — CHANGES_REQUESTED** [review]

File: src/dp/convex_hull_optimization.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (harry900831) [reply src/dp/convex_hull_optimization.md:503]: @aftermaaath So are you going to delete this part? Or you gonna remain the paragraph you wrote. Both are fine for me.

**Pull request merged**